### PR TITLE
Adding delay calculators (delay-utils) to utils.jl

### PR DESCRIPTION
**Pull request by CrocoDuckoDucks**
Hi there!

I wrote few simple methods similar to Matlab's [finddelay](http://uk.mathworks.com/help/signal/ref/finddelay.html?requestedDomain=www.mathworks.com) and [alignsignals](http://uk.mathworks.com/help/signal/ref/alignsignals.html). They are very simple, reflecting that I am Julia newbie. I make massive use of these methods and I though someone else might like them, so here my Pull Request. I have also other methods working with N-dimensional arrays, but [a bug in xcorr](https://github.com/JuliaLang/julia/issues/17351) prevents them to be fast and efficient. Hope you find the methods an interesting suggestion for your project.

### CODE REVIEW

**Comment by jfsantos** [comment]
These are interesting, could you add some tests as well?

**Comment by CrocoDuckoDucks** [comment]
Yep! I have been working on few tests, I just have to polish them up and make them a little more complete. I will commit soon.

**Comment by CrocoDuckoDucks** [comment]
Hi there, I pushed the test I was working on. It is the first time I work with tests, so let me know whether there is something else I should do.

File: src/util.jl [hunk 1]
```diff
@@ -288,4 +293,92 @@ function shiftin!{T}(a::Vector{T}, b::Vector{T})
...
 
+
+# Delay Utility Functions:
+
+# finddelay(): use peak of cross correlation from origin of time to calculate
+# time lag between two Arrays (signals)
+
+function finddelay{T <: Real}(x::AbstractArray{T, 1}, u::AbstractArray{T, 1})
+
+sₓᵤ = xcorr(x, u)
```
Comment (codles) [inline src/util.jl]: The convention in the rest of the package is to indent the content of functions.

File: src/util.jl [hunk 2]
```diff
@@ -288,4 +293,92 @@ function shiftin!{T}(a::Vector{T}, b::Vector{T})
...
+
+function shiftsignals!{T <: Real}(u::AbstractArray{T, 1}, δ::Int)
+
+lᵤ = length(u)
+
+if δ > 0
+
+    deleteat!(u, 1:δ)
+    
+    # append!() could be used, but this is faster and prevents allocation.
```
Comment (codles) [inline src/util.jl]: Nice comments

**Comment by codles** [comment]
Nice tests and comments. Would it be possible to add Inline documentation as discussed in issue #116?

**Comment by CrocoDuckoDucks** [comment]
Yes, I never did it but looks straightforward. I will start working on it as soon as I am back from holidays early next week

**Comment by CrocoDuckoDucks** [comment]
Sorry guys, I underestimated how much of a noob I am. At first I imagine that I was supposed to add documentation directly in src/util.jl following [this](http://docs.julialang.org/en/release-0.4/manual/documentation/). Not sure is that the case tho, as after skimming through the rest of the source code I couldn't see similar lines. I am afraid I have to ask what is the correct way to add inline documentation.

**Comment by jfsantos** [comment]
We currently don't have inline docs for the library (this is the reason #116 is still open!). We currently use the documentation from the `docs` folder to automatically generate the documentation pages at readthedocs.

**Comment by CrocoDuckoDucks** [comment]
Hi there cool people!

It has been a while since I had feedback on this PR, so I was wondering whether I completed it correctly on my side. Maybe I did not mention it, but it is my first PR so I might have missed some step...

**Comment by galenlynch** [comment]
Hi @CrocoDuckoDucks do you have any interest in rebasing this and updating it for Julia 1.0?

**Comment by CrocoDuckoDucks** [comment]
Hi @galenlynch, sorry for the very late reply, I missed all of my notifications for a couple of weeks...

I guess I could have a go. I'll see what I can do and revert back.

**Comment by galenlynch** [comment]
Awesome, thank you! I'll review this soon.

**Comment by CrocoDuckoDucks** [comment]
@galenlynch Good stuff, thanks. I will address the various changes  in the next few days.

**Comment by CrocoDuckoDucks** [comment]
So, I changed a few things.

1. Fixed the typos.
2. Since we prefer to mutate the first argument in mutating functions, I changed all the function signatures so that mutating and not mutating versions are used in the same way.
3. Using `argmax` and `abs` to locate the absolute peak of the cross-correlation.
4. Improved documentation and added examples and references.
5. Removed unicode characters.
6. Removed unnecessary parametric types.
7. Redundancy removed as per suggestions.

It looks like tests are now failing for Julia 0.6. Is this a problem?

**Comment by galenlynch** [comment]
Ok, this PR looks great to me now. Once the tests pass on 0.6, I will be more than happy to merge it.

**Comment by galenlynch** [comment]
@CrocoDuckoDucks you should be able to fix the errors on 0.6 by using `Compat.mapreduce`. Sorry for suggesting something that didn't work!

**Comment by CrocoDuckoDucks** [comment]
@galenlynch No problem, I went for a bit of trial and error. Sorry if that is making a bit of noise with notifications and the like. I am not entirely sure the behaviour of `finddelay` is what I want. I will do some experiments locally and report back.

**Comment by galenlynch** [comment]
No problem! Now that I think about it, I don't really understand why you are taking the absolute value of the correlation in the first place, it seems like only positive lags would be desired. Also, there's another bug in my suggested implementation (finding the minimum lag, instead of the minimum absolute value of lag).

Here's my suggestion with all the bug fixes:
```julia
function finddelay(x::AbstractVector{<: Real}, y::AbstractVector{<: Real})
    s = xcorr(y, x)
    max_corr = maximum(s)
    max_idxs = Compat.findall(x -> x == max_corr, s)

    center_idx = cld(length(s), 2)
    # Delay is position of peak cross-correlation relative to center.
    # If the maximum cross-correlation is not unique, use the position
    # closest to the center.
    d_ind = Compat.argmin(abs.(center_idx .- max_idxs))
    d = center_idx - max_idxs[d_ind]
end

```

**Comment by CrocoDuckoDucks** [comment]
> No problem! Now that I think about it, I don't really understand why you are taking the absolute value of the correlation in the first place, it seems like only positive lags would be desired.

Often one wants to find the delay of some kind of filter, which could have negative gain. If we pick the absolute peak of the cross-correlation we find the delay also in that case. For example:
```julia
julia> finddelay([0, 0, -1, -2, -3], [1, 2, 3])
2
```
So, the delay was correctly calculated also in case of phase inversion. Instead if we peak the maximum:
```julia
julia> finddelay([0, 0, -1, -2, -3], [1, 2, 3])
-1
```

**Comment by CrocoDuckoDucks** [comment]
I committed a revision I am happy with. Let me know what you think.

**Comment by galenlynch** [comment]
It seems you've changed `finddelay` to find the lag with maximum correlation, instead of maximum absolute correlation. From our last conversation, it sounded like you want it to be absolute correlation?

File: src/util.jl [hunk 3]
```diff
@@ -336,4 +341,117 @@ function shiftin!(a::AbstractVector{T}, b::AbstractVector{T}) where T
...
+# Example
+```jldoctest
+julia> finddelay([0, 0, 1, 2, 3], [1, 2, 3])
+2
+
+julia> finddelay([1, 2, 3], [0, 0, 1, 2, 3])
+-2
+```
+"""
+function finddelay(x::AbstractVector{<: Real}, y::AbstractVector{<: Real})
```
Comment (CrocoDuckoDucks) [inline src/util.jl]: @galenlynch the function below is searching for the peak of the absolute correlation.

```julia
s = xcorr(y, x)
max_corr = maximum(abs, s)
max_idxs = Compat.findall(x -> abs(x) == max_corr, s)
```

**Comment by galenlynch** [comment]
Thanks for sticking with it!

**Pull request merged**